File: ts/src/identity-provider-config.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { createRealm } from "./realm.js";
import { RealmError } from "./errors.js";

interface Captured {
  method: string;
  url: string;
  body?: unknown;
}

function mkFetch(handler: (req: Captured) => Response): typeof fetch {
  return (async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input.toString();
    const method = init?.method ?? "GET";
    if (url.endsWith("/auth/login")) {
      return new Response(JSON.stringify({ status: "ok", subject_type: "platform", refresh_token: "rtok-platform", access_token: "pt_x", expires_in: 300 }), {
        status: 200, headers: { "content-type": "application/json" },
      });
    }
    let body: unknown;
    if (typeof init?.body === "string" && init.body.length > 0) {
      try { body = JSON.parse(init.body); } catch { body = init.body; }
    }
    return handler({ method, url, body });
  }) as typeof fetch;
}

const REALM = { realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test" } as const;

const SAMPLE = {
  id: "idp-1", entity_type: "realm", entity_id: "r", provider: "google",
  client_type: "web", client_id: "gid", allowed_origins: ["https://app.example"],
  comments: "", enabled: true, created_at: 1, updated_at: 2,
};

test("idp.list: injects platform_id=realmId and normalizes items", async () => {
  const fetch = mkFetch((req) => {
    assert.equal(req.method, "GET");
    assert.match(req.url, /\/identity-providers\?/);
    assert.match(req.url, /platform_id=r/);
    return new Response(JSON.stringify({ items: [SAMPLE] }), {
      status: 200, headers: { "content-type": "application/json" },
    });
  });
  const realm = createRealm({ ...REALM, fetch });
  const page = await realm.identityProviderConfig.list();
  assert.equal(page.items.length, 1);
  assert.equal(page.items[0]!.provider, "google");
});

test("idp.list: forwards tenant_id and normalizes absent items to []", async () => {
  const fetch = mkFetch((req) => {
    assert.match(req.url, /platform_id=r/);
    assert.match(req.url, /tenant_id=t9/);
    return new Response(JSON.stringify({}), {
      status: 200, headers: { "content-type": "application/json" },
    });
  });
  const realm = createRealm({ ...REALM, fetch });
  const page = await realm.identityProviderConfig.list({ tenantId: "t9" });
  assert.deepEqual(page.items, []);
});

test("idp.create: injects platform_id, maps camel->snake, omits unset", async () => {
  const fetch = mkFetch((req) => {
    assert.equal(req.method, "POST");
    assert.match(req.url, /\/identity-providers$/);
    assert.deepEqual(req.body, {
      platform_id: "r",
      provider: "google",
      client_type: "web",
      client_id: "gid",
      allowed_origins: ["https://app.example"],
    });
    return new Response(JSON.stringify(SAMPLE), {
      status: 201, headers: { "content-type": "application/json" },
    });
  });
  const realm = createRealm({ ...REALM, fetch });
  const out = await realm.identityProviderConfig.create({
    provider: "google", clientType: "web", clientId: "gid",
    allowedOrigins: ["https://app.example"],
  });
  assert.equal(out.id, "idp-1");
});

test("idp.create+update: provider config (Firebase web config) round-trips", async () => {
  const fb = { apiKey: "AIza-test", authDomain: "demo-app.firebaseapp.com", projectId: "demo-app" };

  // create includes config
  let fetch = mkFetch((req) => {
    assert.deepEqual((req.body as Record<string, unknown>)["config"], fb);
    return new Response(JSON.stringify({ ...SAMPLE, provider: "firebase", config: fb }), {
      status: 201, headers: { "content-type": "application/json" },
    });
  });
  let realm = createRealm({ ...REALM, fetch });
  let out = await realm.identityProviderConfig.create({
    provider: "firebase", clientType: "web", clientId: "demo-app",
    allowedOrigins: ["https://app.example.com"], config: fb,
  });
  assert.deepEqual(out.config, fb);

  // patch replaces config wholesale
  fetch = mkFetch((req) => {
    assert.deepEqual(req.body, { config: fb });
    return new Response(JSON.stringify({ ...SAMPLE, config: fb }), {
      status: 200, headers: { "content-type": "application/json" },
    });
  });
  realm = createRealm({ ...REALM, fetch });
  out = await realm.identityProviderConfig.update("idp-1", { config: fb });
  assert.deepEqual(out.config, fb);
});

test("idp.create: 409 provider_exists surfaces as RealmError(conflict)", async () => {
  const fetch = mkFetch(() => new Response(JSON.stringify({
    error: { code: "conflict", message: "provider already configured" },
    provider_exists: true,
  }), { status: 409, headers: { "content-type": "application/json" } }));
  const realm = createRealm({ ...REALM, fetch });
  await assert.rejects(() => realm.identityProviderConfig.create({
    provider: "google", clientType: "ios", clientId: "gid",
  }), (e: Error) => e instanceof RealmError && e.code === "conflict" && e.httpStatus === 409);
});

test("idp.update: sends only provided fields", async () => {
  const fetch = mkFetch((req) => {
    assert.equal(req.method, "PATCH");
    assert.match(req.url, /\/identity-providers\/idp-1$/);
    assert.deepEqual(req.body, { enabled: false, client_id: "new" });
    return new Response(JSON.stringify({ ...SAMPLE, enabled: false, client_id: "new" }), {
      status: 200, headers: { "content-type": "application/json" },
    });
  });
  const realm = createRealm({ ...REALM, fetch });
  const out = await realm.identityProviderConfig.update("idp-1", { enabled: false, clientId: "new" });
  assert.equal(out.enabled, false);
  assert.equal(out.client_id, "new");
});

test("idp.delete: returns deleted ack", async () => {
  const fetch = mkFetch((req) => {
    assert.equal(req.method, "DELETE");
    assert.match(req.url, /\/identity-providers\/idp-1$/);
    return new Response(JSON.stringify({ status: "deleted" }), {
      status: 200, headers: { "content-type": "application/json" },
    });
  });
  const realm = createRealm({ ...REALM, fetch });
  const out = await realm.identityProviderConfig.delete("idp-1");
  assert.equal(out.status, "deleted");
});

test("idp.delete: 404 provider_not_found surfaces as RealmError(not_found)", async () => {
  const fetch = mkFetch(() => new Response(JSON.stringify({
    error: { code: "not_found", message: "provider not found" },
  }), { status: 404, headers: { "content-type": "application/json" } }));
  const realm = createRealm({ ...REALM, fetch });
  await assert.rejects(() => realm.identityProviderConfig.delete("missing"), (e: Error) =>
    e instanceof RealmError && e.code === "not_found" && e.httpStatus === 404);
});

File: go/identity_provider_config_test.go
package realmid

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestIDPConfig_ListInjectsPlatformID(t *testing.T) {
	mux := http.NewServeMux()
	mintPlatformToken(mux)
	mux.HandleFunc("/identity-providers", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "GET" {
			t.Errorf("method=%s", r.Method)
		}
		if pid := r.URL.Query().Get("platform_id"); pid != testRealmID {
			t.Errorf("platform_id=%q want %q", pid, testRealmID)
		}
		if tid := r.URL.Query().Get("tenant_id"); tid != "t-1" {
			t.Errorf("tenant_id=%q", tid)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"items": []any{
				map[string]any{
					"id": "idp-1", "entity_type": "realm", "entity_id": testRealmID,
					"provider": "google", "client_type": "web", "client_id": "g-123",
					"allowed_origins": []string{"https://app.example.com"},
					"enabled":         true, "created_at": 1, "updated_at": 2,
				},
			},
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	page, err := r.IdentityProviderConfig.List(context.Background(), &IDPConfigListOpts{TenantID: "t-1"})
	if err != nil {
		t.Fatalf("list: %v", err)
	}
	if len(page.Items) != 1 || page.Items[0].Provider != "google" {
		t.Errorf("items=%+v", page.Items)
	}
}

func TestIDPConfig_ListNormalizesNilItems(t *testing.T) {
	mux := http.NewServeMux()
	mintPlatformToken(mux)
	mux.HandleFunc("/identity-providers", func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	page, err := r.IdentityProviderConfig.List(context.Background(), nil)
	if err != nil {
		t.Fatalf("list: %v", err)
	}
	if page.Items == nil || len(page.Items) != 0 {
		t.Errorf("items should normalize to empty slice, got %#v", page.Items)
	}
}

func TestIDPConfig_CreateMapsWireShape(t *testing.T) {
	mux := http.NewServeMux()
	mintPlatformToken(mux)
	mux.HandleFunc("/identity-providers", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "POST" {
			t.Errorf("method=%s", r.Method)
		}
		raw, _ := io.ReadAll(r.Body)
		var got map[string]any
		_ = json.Unmarshal(raw, &got)
		if got["platform_id"] != testRealmID {
			t.Errorf("platform_id not injected: %v", got)
		}
		if got["provider"] != "google" || got["client_type"] != "web" || got["client_id"] != "g-123" {
			t.Errorf("body=%v", got)
		}
		origins, ok := got["allowed_origins"].([]any)
		if !ok || len(origins) != 1 {
			t.Errorf("allowed_origins=%v", got["allowed_origins"])
		}
		w.WriteHeader(201)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"id": "idp-1", "entity_type": "realm", "entity_id": testRealmID,
			"provider": "google", "client_type": "web", "client_id": "g-123",
			"allowed_origins": []string{"https://app.example.com"},
			"enabled":         true, "created_at": 1, "updated_at": 1,
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	got, err := r.IdentityProviderConfig.Create(context.Background(), IDPConfigCreate{
		Provider: "google", ClientType: "web", ClientID: "g-123",
		AllowedOrigins: []string{"https://app.example.com"},
	})
	if err != nil {
		t.Fatalf("create: %v", err)
	}
	if got.ID != "idp-1" || !got.Enabled {
		t.Errorf("got=%+v", got)
	}
}

func TestIDPConfig_CreateConflictSurfacesErrIDPExists(t *testing.T) {
	mux := http.NewServeMux()
	mintPlatformToken(mux)
	mux.HandleFunc("/identity-providers", func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(409)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"error": map[string]any{"code": "conflict", "message": "exists"},
			"code":  "provider_exists",
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	_, err := r.IdentityProviderConfig.Create(context.Background(), IDPConfigCreate{
		Provider: "google", ClientType: "web", ClientID: "g-123",
		AllowedOrigins: []string{"https://app.example.com"},
	})
	if err == nil {
		t.Fatal("expected error")
	}
	if !errors.Is(err, ErrIDPExists) {
		t.Errorf("want ErrIDPExists, got %v", err)
	}
	if !IsCode(err, ErrCodeConflict) {
		t.Errorf("want underlying conflict code, got %v", err)
	}
}

func TestIDPConfig_UpdateSendsOnlyProvidedFields(t *testing.T) {
	mux := http.NewServeMux()
	mintPlatformToken(mux)
	mux.HandleFunc("/identity-providers/idp-1", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "PATCH" {
			t.Errorf("method=%s", r.Method)
		}
		raw, _ := io.ReadAll(r.Body)
		var got map[string]any
		_ = json.Unmarshal(raw, &got)
		if _, has := got["client_id"]; has {
			t.Errorf("client_id should be omitted: %v", got)
		}
		if v, ok := got["enabled"].(bool); !ok || v != false {
			t.Errorf("enabled=%v", got["enabled"])
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"id": "idp-1", "entity_type": "realm", "entity_id": testRealmID,
			"provider": "google", "client_type": "web", "client_id": "g-123",
			"enabled": false, "created_at": 1, "updated_at": 5,
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	disabled := false
	got, err := r.IdentityProviderConfig.Update(context.Background(), "idp-1", IDPConfigPatch{Enabled: &disabled})
	if err != nil {
		t.Fatalf("update: %v", err)
	}
	if got.Enabled {
		t.Errorf("expected disabled, got %+v", got)
	}
}

func TestIDPConfig_DeleteHappy(t *testing.T) {
	mux := http.NewServeMux()
	mintPlatformToken(mux)
	mux.HandleFunc("/identity-providers/idp-1", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "DELETE" {
			t.Errorf("method=%s", r.Method)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"status": "deleted"})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	out, err := r.IdentityProviderConfig.Delete(context.Background(), "idp-1")
	if err != nil {
		t.Fatalf("delete: %v", err)
	}
	if out.Status != "deleted" {
		t.Errorf("status=%q", out.Status)
	}
}

func TestIDPConfig_Delete404SurfacesErrIDPNotFound(t *testing.T) {
	mux := http.NewServeMux()
	mintPlatformToken(mux)
	mux.HandleFunc("/identity-providers/missing", func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(404)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"error": map[string]any{"code": "not_found", "message": "no such provider"},
			"code":  "provider_not_found",
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	_, err := r.IdentityProviderConfig.Delete(context.Background(), "missing")
	if err == nil {
		t.Fatal("expected error")
	}
	if !errors.Is(err, ErrIDPNotFound) {
		t.Errorf("want ErrIDPNotFound, got %v", err)
	}
}

// TestIDPConfig_CreateAndPatchSendConfig covers the provider PUBLIC config
// map (e.g. the Firebase web config) on both write paths: present-on-create
// and wholesale-replace-on-patch.
func TestIDPConfig_CreateAndPatchSendConfig(t *testing.T) {
	fb := map[string]string{
		"apiKey":     "AIza-test",
		"authDomain": "demo-app.firebaseapp.com",
		"projectId":  "demo-app",
	}

	t.Run("create includes config", func(t *testing.T) {
		mux := http.NewServeMux()
		mintPlatformToken(mux)
		mux.HandleFunc("/identity-providers", func(w http.ResponseWriter, r *http.Request) {
			raw, _ := io.ReadAll(r.Body)
			var got map[string]any
			_ = json.Unmarshal(raw, &got)
			cfg, ok := got["config"].(map[string]any)
			if !ok || cfg["apiKey"] != "AIza-test" || cfg["authDomain"] != "demo-app.firebaseapp.com" {
				t.Errorf("config not sent on create: %v", got["config"])
			}
			_ = json.NewEncoder(w).Encode(map[string]any{
				"id": "idp-1", "entity_type": "realm", "entity_id": testRealmID,
				"provider": "firebase", "client_type": "web", "client_id": "demo-app",
				"config": fb, "enabled": true, "created_at": 1, "updated_at": 1,
			})
		})
		srv := httptest.NewServer(mux)
		defer srv.Close()

		r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
		got, err := r.IdentityProviderConfig.Create(context.Background(), IDPConfigCreate{
			Provider: "firebase", ClientType: "web", ClientID: "demo-app",
			AllowedOrigins: []string{"https://app.example.com"}, Config: fb,
		})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		if got.Config["projectId"] != "demo-app" {
			t.Errorf("config not parsed back: %+v", got.Config)
		}
	})

	t.Run("patch replaces config", func(t *testing.T) {
		mux := http.NewServeMux()
		mintPlatformToken(mux)
		mux.HandleFunc("/identity-providers/idp-1", func(w http.ResponseWriter, r *http.Request) {
			raw, _ := io.ReadAll(r.Body)
			var got map[string]any
			_ = json.Unmarshal(raw, &got)
			cfg, ok := got["config"].(map[string]any)
			if !ok || cfg["apiKey"] != "AIza-test" {
				t.Errorf("config not sent on patch: %v", got["config"])
			}
			_ = json.NewEncoder(w).Encode(map[string]any{
				"id": "idp-1", "entity_type": "realm", "entity_id": testRealmID,
				"provider": "firebase", "client_type": "web", "client_id": "demo-app",
				"config": fb, "enabled": true, "created_at": 1, "updated_at": 9,
			})
		})
		srv := httptest.NewServer(mux)
		defer srv.Close()

		r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
		got, err := r.IdentityProviderConfig.Update(context.Background(), "idp-1", IDPConfigPatch{Config: &fb})
		if err != nil {
			t.Fatalf("update: %v", err)
		}
		if got.Config["authDomain"] != "demo-app.firebaseapp.com" {
			t.Errorf("config not parsed back: %+v", got.Config)
		}
	})
}
